### ISSUE TITLE
DoNotCache(remote_update_cache) option has logical error.

### ISSUE THREAD

**Issue by danni-yan** [bazelbuild/reclient#87]
rewrapper offered a `remote_update_cache` option. 
https://github.com/bazelbuild/reclient/blob/083a60f2afd8887ff5c8a82630a9700149e9b4fa/cmd/rewrapper/main.go#L88

When passed to reproxy, it become `DoNotCache`. But the `DoNotCache` value is part of Action message: 
https://github.com/bazelbuild/remote-apis-sdks/blob/c08ea7c49ca490067f47bfa8d699178ae37d7b12/go/pkg/rexec/rexec.go#L188

When `ExecutionStrategy=local`  and `DoNotCache=yes`, the CachedResult ONLY can be null. Because the Action digest is diff with  `DoNotCache=false`. 

And after runLocal the cacheLocal judged the DoNotCache option.
https://github.com/bazelbuild/reclient/blob/083a60f2afd8887ff5c8a82630a9700149e9b4fa/internal/pkg/reproxy/action.go#L716

in the execContext.UpdateCachedResult() judged the DoNotCache option again(the sdk did this).
https://github.com/bazelbuild/reclient/blob/083a60f2afd8887ff5c8a82630a9700149e9b4fa/internal/pkg/reproxy/action.go#L728

In summary, with the current code implementation, it is impossible to hit remote cache when `DoNotCache=yes`.

This is bazel source. there're two options: `remoteUploadLocalResults` and `NO_REMOTE_CACHE_UPLOAD`
https://github.com/bazelbuild/bazel/blob/887e1bd1ad860da91226ba67dfca0eebd8754e8c/src/main/java/com/google/devtools/build/lib/remote/util/Utils.java#L576

**Comment by mrahs** [comment]
There is an override in rewrapper to avoid this problem: https://github.com/bazelbuild/reclient/blob/083a60f2afd8887ff5c8a82630a9700149e9b4fa/internal/pkg/rewrapper/rewrapper.go#L183-L188

However, this still means that LERC can only use cache hits from another LERC with the same reclient version.

**Comment by mrahs** [comment]
This is a recent change that seems to force updating the cache for the local strategy in order to benefit from cache hits of other strategies. This could be conflicting with the semantics of the flags and the correctness protection in https://github.com/bazelbuild/reclient/blob/981e9fd32786718254a56b78ee8f2c26de1ff378/internal/pkg/reproxy/server.go#L438

I'll take a deeper look soon to verify.

**Comment by ramymedhat** [comment]
`DoNotCache` is not passed to the remote execution service in the case where the exec strategy is `LOCAL`. The code in https://github.com/bazelbuild/reclient/issues/87#issuecomment-2293709854 unsets it in the **Remote**ExecutionOptions, not the **Local**ExecutionOptions. So the action should still get a cache hit because `DoNotCache` is not part of the action proto.

**Comment by danni-yan** [comment]
yes! My rewrapper is an older version! I need to upgrade it to a version after your patch.
